Check group content entity operation access

pfrenssen opened the pull request:
This is a followup for #192. Now that we can assign default roles to users which are populated with a set of default permissions for doing CRUD operations on group content, we should use these when checking access.

This PR will e.g. allow members to edit nodes in a group if they have the OG permission "edit any node_type node".

pfrenssen committed:
Check access to do CRUD operations on group content.

pfrenssen commented:
This builds on #192 so it contains all that code. The actual implementation is in the last commit https://github.com/amitaibu/og/pull/217/commits/6aca08c59cb9060634eb7991b855215284a2c374.

This still needs tests.

pfrenssen commented on src/OgAccess.php:
```diff
@@ -223,6 +227,83 @@ public static function userAccessEntity($operation, EntityInterface $entity, Acc
+    }
+
+    // @todo Also deal with the use case that CRUD operations are granted to
+    //   non-members.
```
For the moment to get OG permissions we get them from the OgMembership object. It's probably a good idea to make some specific helper methods to retrieve the set of OG permissions for group content for a particular user, whether they are member or not.

pfrenssen commented on src/OgAccess.php:
```diff
@@ -203,6 +203,10 @@ public static function userAccessEntity($operation, EntityInterface $entity, Acc
+          }
           $user_access = static::userAccess($group, $operation, $user);
           if ($user_access->isAllowed()) {
             return $user_access->addCacheTags($cache_tags);
```
I am not 100% sure if we still need this call to `static::userAccess()`. I was not sure of the original use case for this so I haven't touched it. `static::userAccess()` is concerned with groups rather than group content, and the groups have already been dealt with earlier in this method.

Edit: This has in the meantime been answered. The `static::userAccess()` method is intended for checking normal permissions such as `administer group`.

pfrenssen commented on tests/src/Kernel/PermissionEventIntegrationTest.php:
```diff
@@ -0,0 +1,128 @@
+    return [
+      [FALSE, $default_permissions],
+      [TRUE, $default_permissions + $group_content_permissions],
+    ];
```
Just noticed that this test is wrong, we should use `array_merge()` instead of the array union operator `+` for indexed arrays. If I fix this the test fails. The group content permissions are not added :(

pfrenssen commented:
I'm using static methods because I'm following the established pattern, but this is now going to break the unit tests again, since I am statically calling the `PermissionManager` service. This is going to require some trickery again to make the tests pass :(

pfrenssen committed:
Do not use CRUD in the method names, there are more operations than only CRUD operations.

pfrenssen committed:
Reuse the generation of the entity operation permissions.

pfrenssen commented:
I have converted `OgAccess` into a service in #226. This solves the problem with the unit tests.

Merging this was a bit awkward, I'm going to rebase this branch on top of #226 so that the commit progression is more logical.

pfrenssen committed:
Inject the permission manager.

pfrenssen committed:
Use the injected account proxy to get information about the current user.

pfrenssen committed:
Inject the permission manager now that OgAccess is a service.

pfrenssen committed:
Update OgAccessEntityTest now that the entity operations are also checked for access.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

amitaibu commented:
This will require a re-roll

pfrenssen commented:
Picking this back up, will first merge in latest changes and try to get a green test run, then I'll have a look to make a list of what still needs to be done here.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Update getPermissionList() now that the permissions are keyed by entity operation.

pfrenssen committed:
Don't speak about CRUD, entity operations go beyond the basic CRUD operations.

pfrenssen committed:
Update documentation.

pfrenssen commented:
I am thinking how we can best provide the entity operation permissions.

We have two types of permissions:
1. "Group level permissions", such as 'subscribe without approval' and 'manage users'. These are currently provided by the `PermissionEvent` subscribers.
2. "Entity operation level permissions" on group content, so that we can allow members to create and edit group content.

In D7 we only supplied a hard coded set of permissions for the Node module, but in D8's everything-is-an-entity world we should extend this to all entities. A good example use case is the OG Menu module, in which each menu is group content type of the custom `OgMenuInstance` entity. In order to allow certain group roles to create or edit OG Menus we need know that OG Menu exposes this entity type, and that it supports certain operations on it.

In core we don't have this information currently (ref. #222). We need to retrieve it somehow, through a hook or event listener.

I am now thinking of extending the current `PermissionEvent` to handle both the group level and entity operation level permissions. This seems better from a DX perspective, so that modules don't need to implement two subscribers to declare a set of permissions.

The problem with the entity operation permissions in Drupal is that there is no formal framework, and the permissions are based on human language. A good example is the basic 'edit own article content' permission - neither the operation ('update') or the entity type ('node') are even mentioned in the permission, it's impossile to parse the needed information out of it.

So we need to be able to associate a permission with an operation, and with an entity type and bundle. We also need to distinguish between a user having permission to operate on their own content, or on all content: ('edit own article content' vs 'edit any article content'). We also need the machine name and the title.

It seems like it would be a good idea to use a class for this, so something like the 'edit any article content' can be provided in the permission subscriber using a new method like `setEntityOperationPermission()`:

```
  public function provideDefaultOgPermissions(PermissionEventInterface $event) {
    $permission = new EntityOperationPermission();
    $permission
      ->setName('edit any article content')
      ->setTitle(t('Article: edit any content'))
      ->setDescription(t('Allows to edit any node of type article'))
      ->setEntityType('node')
      ->setBundle('article')
      ->setOwnership('any')
      ->setOperation('update');
    $event->setEntityOperationPermission($permission);

    // Here is an example of a normal group-level permission being set in the
    // same event subscriber.
    $event->setPermissions([
      'administer group' => [
        'title' => t('Administer group'),
        'description' => t('Manage group members and content in the group.'),
        'default roles' => [OgRoleInterface::ADMINISTRATOR],
        'restrict access' => TRUE,
      ],
    ]);
  }
```

For basic CRUD operations we can supply all permissions out of the box with generic wording such as `update own $bundle_id $entity_type_id`. We can then allow modules to alter this if the generic wording is not to their liking.

pfrenssen commented:
On second thought, I'm going to use an array structure for the moment to pass the data, to be consistent with the group level permissions. This would make it a lot less work to implement. We can later on still make an object to store the permissions.

If we do it like this we don't even need two separate methods, we can pass everything in the same method, and distinguish between the two if the 'operation' key is present.

So the example implementation above would instead look like this:

```
  public function provideDefaultOgPermissions(PermissionEventInterface $event) {
    $event->setPermissions([
      // Example entity operation type permission.
      'edit any article content' => [
        'title' => t('Article: edit any content'),
        'description' => t('Allows to edit any node of type article'),
        'entity type' => 'node',
        'bundle' => 'article',
        'ownership' => 'any',
        'operation' => 'update',
      ],

      // Example group-level type permission.
      'administer group' => [
        'title' => t('Administer group'),
        'description' => t('Manage group members and content in the group.'),
        'default roles' => [OgRoleInterface::ADMINISTRATOR],
        'restrict access' => TRUE,
      ],
    ]);
  }
```

amitaibu commented:
This should be split into two methods. One is declaring the permissions that might be used.  And the second is the permissions associated with a certain role.

(As you could have a role with no permissions at all)

amitaibu commented:
Also my thoughts for providing default roles is that we maybe should pass a partially applied OgRole instead of the array.

And when we create the roles we just need to fill in the group type and bundle/group I'd

pfrenssen commented:
I'm not entirely sure what you mean. This code is intended to be supplied by modules to declare their own OG permissions, they don't know to which roles these permissions apply.

For the passing of the partially applied OgRole, you mean to do this where the event is called? This is already implemented, it currently looks like this (from `GroupManager::createPerBundleRoles()`):

```
  protected function createPerBundleRoles($entity_type_id, $bundle_id) {
      // Populate the default permissions.
      $event = new PermissionEvent($entity_type_id, $bundle_id);
      /** @var \Drupal\og\Event\PermissionEventInterface $permissions */
      $permissions = $this->eventDispatcher->dispatch(PermissionEventInterface::EVENT_NAME, $event);
      $properties['permissions'] = array_keys($permissions->filterByDefaultRole($role_name));

      $role = $this->ogRoleStorage->create($properties + $default_properties);
      $role->save();
    }
  }
```

We could probably create the `OgRole` first and then put it in the `PermissionEvent` and apply the permissions to the roles inside the `PermissionEvent`, but to me it seems like it's not the responsibility of the `PermissionEvent`.

I would try to avoid a direct hard coupling between `OgRole` and `PermissionEvent`, but I agree that the applying of the permissions to the role is also not something that belongs in `GroupManager` either. The natural place to handle this seems `PermissionManager`.

pfrenssen commented:
If I'm dreaming that we are doing this 100% right we shouldn't be messing with arrays at all :) We could make a proper `OgPermission` object that has all the relevant getters and setters, and can do validation internally. Then we can set the permissions on the role like this:

```
  protected function createPerBundleRoles($entity_type_id, $bundle_id) {
    foreach ($this->getDefaultRoles() as $role_name => $default_properties) {
      $properties = [
        'group_type' => $entity_type_id,
        'group_bundle' => $bundle_id,
        'id' => $role_name,
        'role_type' => OgRole::getRoleTypeByName($role_name),
        'permissions' => array_keys(PermissionManager::getDefaultPermissions($entity_type_id, $bundle_id)),
      ];

      $role = $this->ogRoleStorage->create($properties + $default_properties);
      $role->save();
    }
  }
```

And `PermissionManager::getDefaultPermissions()` would do the heavy lifting, returning an array of `OgPermission` objects, keyed by permission name.

In this case we only need the permission names, but the full objects will be handy when doing things like building the permission matrix in the `og_ui` module.

pfrenssen committed:
Put the additional metadata in the permission array.

pfrenssen committed:
If we handle all of this through the PermissionEvent we don't need to be able to alter it any more. It can already be altered through the event.

pfrenssen committed:
Check the presence of the group content entity type and bundle for operations.

pfrenssen committed:
WIP documentation.

amitaibu commented:
What I think we can do is something like:

``` php
protected function createPerBundleRoles($entity_type_id, $bundle_id) {

  foreach ($this->getDefaultRoles() as $og_role) {
    // Get the permissions by the role name.
    $permissions = PermissionManager::getDefaultPermissions($og_role->getLabel());

    /** @var \Drupal\og\Entity\OgRoleInterface $og_role */
    $og_role
      ->setGroupType($group_type)
      ->setId($bundle_id)
      // ...
      ->setPermissions($permissions),
      ->save();
  }
```

As reference you can follow the flow of OG7's `og_roles_override()` where it first called to get the default roles and then get the default permissions that should be assigned to the role (by the role name)

amitaibu commented:
(btw, Passing the partially applied OgRole (i.e. one that doesn't have the group type/ bundle set yet) would also fix #232 )

pfrenssen commented:
OK that's very clear, I like this approach, it lets `PermissionManager` handle all the logic, including the filtering which is now handled in `PermissionEvent`. That will reduce `PermissionEvent` to the simple storage class it is intended to be.

I'll also try to sneak in the fix for #232.

pfrenssen committed:
Deliberately introduce a failure as a reminder.

pfrenssen committed:
Inject the permission manager. We'll need it.

pfrenssen committed:
Update documentation.

pfrenssen commented:
In order to limit the scope of this I'm going to put the returning of partial OgRoles in a separate PR. That will fix #232 and #216 too.

Once that's done I'll return to this ticket.

pfrenssen commented:
The returning of partially applied `OgRole` objects has been proposed in #237. I'm going to merge that code in here.

The next step will be to refactor the filtering of permissions by default role so that it is no longer split between `GroupManager` and `DefaultPermissionEvent` but is handled by `PermissionManager` where it belongs. I will also do this in a separate ticket to limit the scope and make it easier to review.

pfrenssen committed:
Merge branch 'return-partial-og-roles' into check-group-content-crud-access

pfrenssen commented:
The filtering is now in `PermissionManager` in PR https://github.com/amitaibu/og/pull/238.

Next up is the splitting up of permissions in group-level permissions and group-content-level permissions.

@amitaibu suggested in https://github.com/amitaibu/og/pull/217#issuecomment-225601360 to split this into two different methods:

> This should be split into two methods. One is declaring the permissions that might be used. And the second is the permissions associated with a certain role.

But I've been thinking about this while working on the related tickets in the last couple of days and I think it would make more sense to keep 1 set of methods, and instead pass different 'permission types'. After all these are different data types that require different properties.

So instead of having two separate methods:

```
  public function provideDefaultOgPermissions(PermissionEventInterface $event) {
    $event->setGroupContentPermissions([
      // Example permission for an entity operation on group content.
      'edit any article content' => [
        'title' => t('Article: edit any content'),
        'description' => t('Allows to edit any node of type article'),
        'entity type' => 'node',
        'bundle' => 'article',
        'ownership' => 'any',
        'operation' => 'update',
      ],
    ]);

    $event->setGroupPermissions([
      // Example group-level type permission.
      'administer group' => [
        'title' => t('Administer group'),
        'description' => t('Manage group members and content in the group.'),
        'default roles' => [OgRoleInterface::ADMINISTRATOR],
        'restrict access' => TRUE,
      ],
    ]);
  }
```

We could do something like this:

```
  public function provideDefaultOgPermissions(PermissionEventInterface $event) {
    $event->setPermissions([
      // Example permission for an entity operation on group content.
      new OgGroupContentPermission([
        'name' => 'edit any article content',
        'title' => t('Article: edit any content'),
        'description' => t('Allows to edit any node of type article'),
        'entity type' => 'node',
        'bundle' => 'article',
        'ownership' => 'any',
        'operation' => 'update',
      ]),

      // Example group-level type permission.
      new OgGroupPermission([
        'name' => 'administer group',
        'title' => t('Administer group'),
        'description' => t('Manage group members and content in the group.'),
        'default roles' => [OgRoleInterface::ADMINISTRATOR],
        'restrict access' => TRUE,
      ]),
    ]);
  }
```

This would be great from a DX standpoint, since we can then provide convenience methods, and easily alter the existing permissions:

```
  public function alterExistingOgPermissions(PermissionEventInterface $event) {
    $permission = $event->getGroupContentPermission('node', 'article', 'update');
    $permission->setDescription(t('Some new help text.'));
    $event->setPermission($permission);
  }
```

pfrenssen commented:
Started working on the splitting of the permissions in #240.

pfrenssen commented:
This is now missing a full week of work (#238 and #240) and everything that has been merged into 8.x-1.x in the meanwhile. Going to attempt a merge to bring this back up to date.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Merge branch 'split-permissions' into check-group-content-crud-access

pfrenssen committed:
WIP

pfrenssen committed:
Fix typo.

pfrenssen committed:
Fix matching of OgRole ID, it failed because 'non-member' contains a dash.

pfrenssen committed:
Musing about future improvements.

pfrenssen committed:
Inject the GroupManager.

pfrenssen committed:
We have PermissionManagerInterface now.

pfrenssen committed:
Temporarily remove early abort.

pfrenssen committed:
Pass on the group content bundle IDs.

pfrenssen committed:
Check if the user has the entity operation permission for the group membership.

pfrenssen committed:
Merge branch 'split-permissions' into check-group-content-crud-access

pfrenssen committed:
Update documentation.

pfrenssen committed:
Add support for retrieving permissions from non-members.

pfrenssen committed:
Small restructuring to make the code more readable.

pfrenssen commented:
I'm going to finish this now. This will probably be the last issue I can work on unfortunately. Our development focus will move away from OG when this is in, since we will have all functionality available that our project needs.

amitaibu commented:
Well, all your efforts are **greatly** appreciated!

pfrenssen commented:
You're welcome! I really love working on OG. I'll still check in from time to time, and I want to finish the roles & permission UI as well, but I'll have to do this in my spare time. And I have a lot of projects that are competing to get a slice of my spare time :)

pfrenssen commented:
I have to fix the failures in `OgAccessEntityTestBase` here, but I cannot understand it in the current form. It is using things like `$this->entity` without any documentation, and it is unclear what this entity contains: it might be a group, or group content?

This was researched and clarified in #230, I will merge in that branch so I can at least make sense of it.

pfrenssen committed:
Merge remote-tracking branch 'origin/cleanup-og-access-test' into check-group-content-crud-access

pfrenssen committed:
More test greenage.

pfrenssen committed:
Access the storage handler directly.

This will be easier to refactor once we turn Og into a service, and it
saves us from having to mock the deprecated EntityManager.

pfrenssen committed:
Move the code to always return Neutral.

It's unclear to me why it is necessary that unsaved groups return neutral for
every possible operation, but as this use case has a dedicated test it appears
to be important.

pfrenssen committed:
Clarify reminder.

pfrenssen committed:
Small optimization pointed out by the unit test coverage.

pfrenssen committed:
Actually use the GroupManager, it has been injected for a reason.

pfrenssen commented:
I need to add a test for this, and I would ideally like to create a unit test, but I have been looking at it for 30 minutes and it really doesn't seem practical at the moment. The number of static methods involved in the process is just too much, it's completely counterproductive. This PR only adds very little code, but I already spent hours yesterday jumping through hoops to get the existing test coverage green.

A very simple section such as this is incredibly problematic for unit testing because it contains a static method call:

```
if ($membership = Og::getMembership($user, $group)) {
  foreach ($membership->getRoles() as $role) {
    // ...
  }
}
```

Because we cannot mock static methods, we now need to mock the entire logic that is happening inside of the method. I started writing this and I got to this point before giving up:

```
// Mock the group entity that is accessed in Og::getMembership().
$entity_type = $this->prophesize(EntityTypeInterface::class);
$entity_type->id()->willReturn($this->entityTypeId);
$group_entity = $this->prophesize(EntityInterface::class);
$group_entity->getEntityTypeId()->willReturn($this->entityTypeId);
$group_entity->bundle()->willReturn($this->bundle);
$group_entity->id()->willReturn($this->groupId);
$this->addCache($group_entity);

// Mock the creation of an empty OgMembership entity, in case the
// permissions are requested for a user that is not a member of the group.
// For non-members the entity that is returned will have the 'non-member'
// role and the 'pending' state.
$non_membership = $this->prophesize(OgMembership::class);
$non_membership->setUser($this->user->reveal())->willReturn($non_membership->reveal());
$non_membership->setEntityId($this->groupId)->willReturn($non_membership->reveal());
$non_membership->setGroupEntityType($this->entityTypeId)->willReturn($non_membership->reveal());
$non_membership->setState(OgMembership::STATE_PENDING)->willReturn($non_membership->reveal());
$non_member_role_id = implode('-', [
  $this->entityTypeId,
  $this->bundle,
  'non-member',
]);
$non_membership->setRoles([$non_member_role_id])->willReturn($non_membership->reveal());
$entity_type_manager->getStorage('og_membership')->willReturn($storage->reveal());
$storage->create(['type' => OgMembership::TYPE_DEFAULT])->willReturn($non_membership->reveal());
$non_member_role = $this->prophesize(OgRole::class);
$non_membership->getRoles()->willReturn($non_member_role);
```

This is not only bad from the standpoint of how much work is involved, but we are also testing the underlying implementation, and this is very brittle, touch one line of code and you're looking at rewriting dozens of lines of test code.

Imagine instead that this was written using a service with dependency injection - the calling code would be incredibly similar:

```
if ($membership = $this->membershipManager()->getMembership($user, $group)) {
  foreach ($membership->getRoles() as $role) {
    // ...
  }
}
```

And the unit testing would be a dream, instead of mimicking dozens of lines of code, just mock the method call and it return value and you're done:

```
$non_member_role = $this->prophesize(OgRole::class);
$non_membership = $this->prophesize(OgMembership::class);
$non_membership->getRoles()->willReturn($non_member_role);
$this->membershipManager()->getMembership($this->user, $this->group)->willReturn($non_membership);
```

In conclusion, I think we should probably not attempt to add any more unit tests for classes that do not consist of 100% OO code. The effort involved and the future maintenance cost is too high. The nice thing is that converting the existing code to OO is not very difficult. We did this for example in #226 and it was relatively quick.

So for this PR I will probably just provide a KernelTest.

pfrenssen committed:
Correct documentation, this was never used as group content.

pfrenssen committed:
Remove unused use statement.

pfrenssen committed:
Start working on the test.

pfrenssen committed:
First round of failure fixing.

amitaibu commented:
Yeah, no need to go nuts over the tests. It's probably the right time as you say to start moving classes to be proper services - which classes do you think we should start with?

pfrenssen committed:
Change ownership from a string to a boolean.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Rename [g|s]etRoles() to [g|s]etApplicableRoles().

pfrenssen committed:
Fix fatal errors discovered while testing.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Working on test coverage.

pfrenssen committed:
Fix bug discovered by test.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen commented:
It seems like my last comment here has not come through. I had answered the question from @amitai and explained what was left to be done here. I probably forgot to submit it, or was possibly on the move with a bad connection.

In short, I have no preference for a certain class to start with. If it was me I'd tackle `Og` first, since it is the biggest one.

What still needs to be done here is to add a functional test that tries to perform an entity operation and checks if the user has access or not. This will cover the last untested part: `og_entity_access()`.

pfrenssen commented on src/Entity/OgRole.php:
```diff
@@ -205,8 +205,8 @@ public function save() {
-      list($entity_type_id, $bundle_id, $name) = explode('-', $this->id());
-      if ($entity_type_id !== $this->getGroupType() || $bundle_id !== $this->getGroupBundle() || $name !== $this->getName()) {
+      $pattern = preg_quote("{$this->getGroupType()}-{$this->getGroupBundle()}-{$this->getName()}");
+      if (!preg_match("/$pattern/", $this->id())) {
```
This fix covers having entity and bundle IDs containing dashes.

pfrenssen commented on src/Entity/OgRole.php:
```diff
@@ -254,7 +254,8 @@ public function set($property_name, $value) {
-    if ($is_locked_property && !$this->isNew()) {
+    $value_has_changed = !empty($original_value = $this->get($property_name)) && $original_value != $value;
+    if ($is_locked_property && !$this->isNew() && $value_has_changed) {
       throw new OgRoleException("The $property_name cannot be changed.");
```
This exception was being thrown during tests, turned it was thrown when an existing role was resaved without changing anything :)

amitaibu replied:
👍

pfrenssen commented on src/Og.php:
```diff
@@ -267,16 +268,32 @@ public static function getMemberships(AccountInterface $user, array $states = [O
+      ->setEntityId($group->id())
+      ->setGroupEntityType($group->getEntityTypeId())
+      ->setState(OgMembershipInterface::STATE_PENDING)
+      ->setRoles([$role_id]);
```
Just saw that this is ignoring `$states` - if `$states` equals `[OgMembershipInterface::STATE_ACTIVE]` then this should still return `NULL` instead of a membership in which the state equals `OgMembershipInterface::STATE_PENDING`.

pfrenssen replied:
This has been fixed, where "fixed" === "this whole section has been removed" :)

pfrenssen commented on src/OgAccess.php:
```diff
@@ -184,16 +206,16 @@ public function userAccess(EntityInterface $group, $operation, AccountInterface
+      // Entity isn't saved yet.
+      if ($entity->isNew()) {
+        return $result->addCacheableDependency($entity);
+      }
```
This is moved in here to satisfy the tests. There is a test specifically added to test that `AccessResult::neutral()` is returned when a new group is passed in. The reason for this is not entirely clear to me, it's not documented here nor in the test, but since there's a test for it I'm assuming that this is important for some reason. I have not investigated any further since it's out of scope for this PR.

amitaibu replied:
> There is a test specifically added to test that AccessResult::neutral() is returned when a new group is passed in.

Can you link to that test, maybe I could clarify.

pfrenssen replied:
Here it is, it has no information about why this is important: https://github.com/amitaibu/og/blob/8.x-1.x/tests/src/Unit/OgAccessEntityTest.php#L39

amitaibu replied:
That test should be removed, it is wrong. We don't care about non saved groups.

Btw, you can also see I have done some fixes to userAccess in #268

pfrenssen replied:
Ok I have removed that test and the corresponding lines that were forcing an `AccessResult::neutral()` on unsaved groups.

pfrenssen committed:
Only return a membership entity in which the user is a non-member if requested.

pfrenssen committed:
Request both the memberships entities of members and non-members to determine if they have access.

amitai commented:
I think you may have tagged the wrong user. 

Sent from my iPad 

On Jul 14, 2016, at 05:58, Pieter Frenssen notifications@github.com wrote:

It seems like my last comment here has not come through. I had answered the question from @amitai and explained what was left to be done here. I probably forgot to submit it, or was possibly on the move with a bad connection.

In short, I have no preference for a certain class to start with. If it was me I'd tackle Og first, since it is the biggest one.

What still needs to be done here is to add a functional test that tries to perform an entity operation and checks if the user has access or not. This will cover the last untested part: {{og_entity_access()}}.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

amitaibu commented:
Hi @amitai nice to meet you, I'm @amitaibu ✋

amitai commented:
Hi! 🙌

 -DABL (mobile)

On Jul 14, 2016, at 7:52 AM, Amitai Burstein notifications@github.com wrote:

Hi @amitai nice to meet you, I'm @amitaibu ✋

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

pfrenssen commented:
Oops! I messed up the usernames, sorry!

amitai commented:
​no problem ;-)

pfrenssen committed:
Add a functional test that checks access to an entity operation through the UI.

pfrenssen commented:
I added a functional test that checks access to group content by visiting `node/ID/edit` with the browser. This validates that the code is invoked correctly from the entity access handler, through `og_entity_access()`.

pfrenssen commented:
This is ready for review if the tests come back green!

amitaibu commented:
Sorry, hectic days. I hope to get to it tomorrow..

pfrenssen commented:
No worries!

amitaibu commented on src/Og.php:
```diff
@@ -268,15 +269,38 @@ public static function getMemberships(AccountInterface $user, array $states = [O
    * @return \Drupal\og\Entity\OgMembership|NULL
-   *   The OgMembership entity, or NULL if the user is not a member of the
-   *   group.
+   *   The OgMembership entity. If the user is not a member of the group, an
```
why this change?

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+   *
+   * @see \Drupal\og\PermissionManager::getEntityOperationPermissions()
+   */
+  public function userAccessGroupContentEntityOperations($operation, EntityInterface $group_entity, EntityInterface $group_content_entity, AccountInterface $user = NULL) {
```
maybe `userAccessGroupContentEntityOperations` => `userAccessGroupContentEntityOperation` (i.e. operation in singular)

pfrenssen replied:
Fixed!

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
   /**
+   * Checks access for entity operations on group content entities.
+   *
+   * This checks if the user has permission to perform the requested operation
```
👍

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+   *   The group content entity for which access to the entity operation is
+   *   requested.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user for which to check access.
```
We should to indicate it's optional, and if empty, the current user will be used.

pfrenssen replied:
Fixed!

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+   */
+  public function userAccessGroupContentEntityOperations($operation, EntityInterface $group_entity, EntityInterface $group_content_entity, AccountInterface $user = NULL) {
+    // Default to the current user.
+    if (!isset($user)) {
```
`$user = $user?: $this->accountProxy->getAccount();`

pfrenssen replied:
:+1:

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+    // Check if the user owns the entity which is being operated on.
+    $is_owner = $group_content_entity instanceof EntityOwnerInterface && $group_content_entity->getOwnerId() == $user->id();
+
+    // Retrieve the permissions and check if our operation is supported.
```
what do you mean by `operation is supported`?

pfrenssen replied:
This wasn't clear indeed. This part is just retrieving the permissions from the permission manager.

pfrenssen replied:
This is fixed.

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+    $permissions = $this->permissionManager->getDefaultEntityOperationPermissions($group_entity_type_id, $group_bundle_id, $group_content_bundle_ids);
+
+    // Filter the permissions by operation and ownership.
+    $ownerships = $is_owner ? [FALSE, TRUE] : [FALSE];
```
Sorry, I'm not clear on those arrays?

pfrenssen replied:
Ah yes, this little piece of magic :)

I tried to clarify it, can you check if this makes more sense now? I couldn't find the right words.

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+      return $permission->getOperation() === $operation && in_array($permission->getOwner(), $ownerships);
+    });
+
+    // @todo This doesn't really vary by user but by the user's role inside of
```
Great idea. Yeah, we should mimic the way core does it for site-wide roles, but on the group level. Can be as a follow up of course.

pfrenssen replied:
Ok I created a followup and referred to it here: #219.

amitaibu commented on src/OgAccess.php:
```diff
@@ -239,6 +270,75 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+    }
+
+    if ($permissions) {
+      // Non-members might also have certain permissions within a group. Request
```
this is probably the reason you return OgMembership for non-members, right?

I think we should look at it a bit differently, as the are a few more cases:
1. User is member (`STATE_ACTIVE`)
2. User is a non-member that has request to get membership and is now pending (`STATE_PENDING`) (i.e. they have a _saved_ `OgMembership`)
3. User is a non-member that has no relation at all to the group (i.e. no `OgMembership`)
4. User is blocked from the group (`STATE_BLOCKED` with a _saved_ `OgMembership`)

That's why I think we need to:
1. Explicitly deny access if user is blocked.
2. While the result might be similar, treat differently non-member who are pending VS just non-members. A use case can be for example, preventing pending non-members from re-requesting group access.

pfrenssen replied:
Yes, there are use cases that you need to be able to grant certain permissions to non-members. For example you might want to allow non-members to create certain group content such as comments. These permissions also need to be stored in an `OgMembership` entity. In case there are no special permissions for non-members, there is no need to store any data so we can generate an empty `OgMembership` on the fly.

I haven't given much thought to blocked users. I can imagine use cases where a blocked user has certain permissions - for example you could have an 'appeal' operation where a blocked user could ask to be unblocked. It is fine for me though to explicitly block all interaction for blocked users. Maybe our only responsibility would be to provide the access to be altered, so sites can still implement their own special behaviour for blocked users. I'll have a look at this again when I pick up the UI counterpart of this.

Your example about preventing pending members from re-requesting membership also makes sense. We must indeed distinguish non-members from pending members.  I will introduce a new `STATE_NON_MEMBER` state for this. I'll also change the constants from numbers to strings, that will improve the DX, e.g. when exporting config you don't have to look up that state "2" means "pending".

Will start on this shortly, going to address the other low hanging fruit first. Github has this annoying tendency of hiding comments as soon as a single context line is touched :)

amitaibu replied:
> These permissions also need to be stored in an OgMembership entity.

As commented above, I don't think an OgMembership is needed for a non-member. Keeping the site-wide vs group analogy, a non-member is like an anonymous user - you just have permissions that are related to anonymous users, but they are not "real" users.

amitaibu replied:
To that matter, when it comes to permissions we treat a `pending` membership like a non-member (with a small exception #283)

amitaibu replied:
> for example you could have an 'appeal' operation where a blocked user could ask to be unblocked

It's a nice use case, but I'd say that for now lets go with the "simple" case and be strict. If you were a jerk and got blocked - you don't get to interact again with the group. Again, using analogy from site-wide vs group: If you got blocked from the site-wide, you can't do anything anymore.

amitaibu replied:
> Your example about preventing pending members from re-requesting membership also makes sense.
> I will introduce a new STATE_NON_MEMBER state for this.

I think you don't need to do this. When it comes to re-subscribe for me the permission is still "allow subscribe". We're just extending the logic, and in out `SubscribeController` can redirect them.

Anyway, by now I think you already got that point that I strongly think we should not create a membership for a non-member 😉

amitaibu replied:
>  I'll also change the constants from numbers to strings, that will improve the DX

👍

pfrenssen replied:
Agree on all points, thanks for clarifying!

I've only changed the constants to strings now.

pfrenssen replied:
> That's why I think we need to:
> 1. Explicitly deny access if user is blocked.

Working on this now. I've started by adding a test. It was expecting that this should already work, since in `OgAccess::userAccessGroupContentEntityOperation()` we are requesting only memberships that have `STATE_ACTIVE` or `STATE_PENDING`. Blocked users are excluded so I assumed this would default to denying access.

Unfortunately the test fails so there seems to be a problem somewhere in the logic. Investigating now.

pfrenssen replied:
Fixed it. Now always returning `FALSE` in `OgMembership::hasPermission()` if the user is blocked.

amitaibu commented on tests/src/Functional/AccessTest.php:
```diff
@@ -0,0 +1,153 @@
+      $this->drupalLogin($this->users[$user]);
+      foreach ($expected_results as $group_content => $expected_result) {
+        $this->drupalGet($this->groupContent[$group_content]->toUrl('edit-form'));
+        $expected_response = $expected_result ? 200 : 403;
```
I think we should go with Kernel test. I think this is the only "browser" test, which we could convert to an  API check to the entity. For me it confirms that from the API level we are getting the access right.

amitaibu commented on tests/src/Functional/AccessTest.php:
```diff
@@ -0,0 +1,153 @@
+      // The member should only have the right to edit their own group content.
+      'member' => ['owner' => FALSE, 'member' => TRUE],
+      // The non-member cannot edit any group content.
+      'non-member' => ['owner' => FALSE, 'member' => FALSE],
```
It would be good to add another scenario, we could call:

``` php
'wiki-non-member' => ['owner' => FALSE, 'member' => FALSE, 'allow_edit' => TRUE],
```

As the name implies, there are use case that a group would - a la wiki - allow non-members (but not blocked ones) to edit their content. (This is were OG really takes access to the far ends of use cases 😉 )

pfrenssen replied:
This is done, I added a `WikiTest` that proves even non-members can be granted entity operation permissions.

amitaibu commented on tests/src/Functional/AccessTest.php:
```diff
@@ -0,0 +1,153 @@
+    parent::setUp();
+
+    // Create a number of test users.
+    $users = [
```
maybe `$users` => `$membership_types`? When I see `$user` I think about `AccountInterface`

pfrenssen replied:
Fixed.

amitaibu commented on tests/src/Functional/AccessTest.php:
```diff
@@ -0,0 +1,153 @@
+
+    // Create two group content items, one owned by the group owner, and one by
+    // the member.
+    foreach (['owner', 'member'] as $user) {
```
(also here $user sounds a bit off)

pfrenssen replied:
Fixed.

amitaibu commented on tests/src/Kernel/Access/OgGroupContentOperationAccessTest.php:
```diff
@@ -0,0 +1,357 @@
+   * An array of test roles.
+   *
+   * @var \Drupal\og\Entity\OgRole[]
+   *   Note that we're not using OgRoleInterface because of a class inheritance
```
👍

amitaibu commented on tests/src/Kernel/Access/OgGroupContentOperationAccessTest.php:
```diff
@@ -0,0 +1,357 @@
+
+    // Create a user that will serve as the group owner. There are no special
+    // permissions granted to the group owner, so this user can be used for
+    // creating entities that are not owned by the user under test.
```
> There are no special permissions granted to the group owner

just as a note, this will be eventually like in D7, controlled by a variable (e.g. `Group owner has full access`)

pfrenssen replied:
That would be very handy! I made an issue for this so this will not be forgotten: https://github.com/amitaibu/og/issues/285

amitaibu commented on tests/src/Kernel/Access/OgGroupContentOperationAccessTest.php:
```diff
@@ -0,0 +1,357 @@
+
+    // Create 3 test roles with associated permissions. We will simulate a
+    // project that has two group content types:
+    // - 'newsletter_subscription': Any registered user can create entities of
```
for brevity, maybe lets use a shorter form `newsletter`

pfrenssen replied:
Great suggestion, fixed!

amitaibu commented on tests/src/Kernel/Access/OgGroupContentOperationAccessTest.php:
```diff
@@ -0,0 +1,357 @@
+    ];
+
+    foreach ($permission_matrix as $role_name => $permissions) {
+      $role_id = "{$this->group->getEntityTypeId()}-{$this->group->bundle()}-$role_name";
```
What do you say about a helper function, so you don't need to know the naming internals. Something like:

`$this->roles[$role_name] = OgRole::loadByGroupAndRoleId($this->group, $role_id);`

pfrenssen replied:
I created this method. I called it `OgRole::loadByGroupAndName()` to be consistent with the other methods that allow to use the role name, such as `OgRole::setName()`.

We had a similar helper function before: `Og::getRole()` but this took the entity type ID, bundle and role name as parameters. I removed it because this one makes a lot more sense: it is in `OgRole` where it belongs, and allowing to pass in a group entity is much better DX than having to extract the entity ID and bundle on the calling side.

I grepped the entire codebase to find where we were using this pattern and replaced it with the new method.

amitaibu replied:
I don't think we need to remove og::getRole completely as you don't necessarily have a group before you want to set the permissions.  But we can of course move it away from Og into OgRole

pfrenssen replied:
`OgRole::loadByGroupTypeBundleAndRoleName()`? Bit long :/

Or maybe `OgRole::getRoleId($group_entity_type, $group_bundle, $role_name)` would be good? Bit less interesting from a DX perspective perhaps. Even though I wouldn't really mind seeing this:

```
$role_id = OgRole::getRoleId($entity_type, $bundle, $role_name);
$role = OgRole::load($role_id);
```

amitaibu commented on tests/src/Kernel/Access/OgGroupContentOperationAccessTest.php:
```diff
@@ -0,0 +1,357 @@
+      }
+    }
+
+    // Create a 'newsletter_subscription' group content type. We are using the
```
nice!

amitaibu commented on tests/src/Kernel/Access/OgGroupContentOperationAccessTest.php:
```diff
@@ -0,0 +1,357 @@
+        $this->groupContent[$bundle_id][$user_id] = $entity;
+      }
+    }
+  }
```
crazy setup! 😃

amitaibu commented:
Wonderful, thank you. I've added a few comments. Most important comment is around handeling OgMembership for non-members.

pfrenssen commented:
Thanks a lot for the great suggestions :) I'll get to it after the weekend.

amitaibu commented:
Needs a re-roll as-well

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Declare visibility on method.

pfrenssen committed:
Argument order of Og::getMembership() has changed.

pfrenssen committed:
Use the new DX friendly way to set the user and group when creating a membership.

pfrenssen committed:
Fix PHP CodeSniffer warnings.

pfrenssen commented:
Getting a false positive for the coding standards check:

```
| ERROR   | [ ] Expected "\Drupal\og\Entity\OgMembership|null"
|         |     but found "\Drupal\og\Entity\OgMembership|NULL"
|         |     for function return type
```

amitaibu commented:
yeah, it should be `null` (lower case)

pfrenssen committed:
Actually is null.

pfrenssen committed:
Use the fullly qualified namespace for return values.

pfrenssen committed:
Use shorthand array syntax.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Rename method. It only deals with a single operation at a time.

pfrenssen committed:
Improve documentation.

pfrenssen committed:
Use the ternary operator instead of an if statement.

pfrenssen committed:
Improve documentation.

pfrenssen committed:
Rename $user to $membership_type since this better conveys what the variable holds.

pfrenssen committed:
Rename 'newsletter_subscription' to 'newsletter'. Shorter is gooder.

pfrenssen committed:
Provide an issue number with a @todo so it becomes actionable.

pfrenssen committed:
Adopt the awesome new DX improvements for creating memberships.

pfrenssen committed:
Remove unused use statement.

amitaibu commented on src/Og.php:
```diff
@@ -249,15 +249,44 @@ public static function getMemberships(AccountInterface $user, array $states = [O
+
+    // The user is not a member, return an OgMembership entity in which the user
+    // has the 'non-member' role.
+    if (in_array(OgMembershipInterface::STATE_PENDING, $states)) {
```
Sorry, I'm not a big fan of this change, as it will require extra checks every time we're doing `Og::getMembership`, to see that the return value has a membersihp with a reference to a role that shouldn't be saved like this (in fact we should probably have some code to defend us from creating memberships referencing a `non-member` role).

The workflow as I see it, still remains easy if you care about creating memberships, e.g.:

``` php
$membership = Og::getMembership($group, $user) ?: Og::createMembership($group, $user);
```

pfrenssen replied:
OK! We can indeed shift this to the calling side.

pfrenssen replied:
Fixed!

pfrenssen replied:
> in fact we should probably have some code to defend us from creating memberships referencing a non-member role

I made a followup for this: https://github.com/amitaibu/og/issues/293

amitaibu commented on src/OgAccess.php:
```diff
@@ -253,6 +284,80 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+
+    $permissions = $this->permissionManager->getDefaultEntityOperationPermissions($group_entity_type_id, $group_bundle_id, $group_content_bundle_ids);
+
+    // Filter the permissions by operation and ownership.
```
👍

amitaibu commented on tests/src/Functional/AccessTest.php:
```diff
@@ -0,0 +1,151 @@
+ *
+ * @group og
+ */
+class AccessTest extends BrowserTestBase {
```
btw, can't Kernel do similar checks? `Functional` is sooo D7 slow ;)

amitaibu replied:
oh, sorry - I see there's also a Kernel. Bonus points!!

pfrenssen replied:
I'll change it to a KernelTest. I just wanted to validate that the entire chain from request to response is working, since access checks are so important. But we can indeed trust that core handles access correctly on their end, and we should only care about our own behaviour.

pfrenssen committed:
Change the membership state constants to strings.

This makes it clearer what each state represents in exported config.

pfrenssen committed:
Small documentation update.

pfrenssen committed:
Test that blocked users cannot create, update or delete any group content.

pfrenssen committed:
Fix PHP CodeSniffer warning.

pfrenssen committed:
Update the membership state field definition, it has changed from integer to string values.

pfrenssen committed:
Move the generation of the 'non-member membership' to the calling side.

pfrenssen committed:
Blocked users should not have any permissions.

pfrenssen committed:
We don't care if a group is new when doing access checks.

pfrenssen committed:
Convert the functional access test into a kernel test.

pfrenssen committed:
Fix PHP CodeSniffer warning.

pfrenssen committed:
Remove unused trait.

pfrenssen committed:
Add a test that proves that it is possible to grant permissions to non-members.

pfrenssen committed:
Provide a convenient method OgRole::loadByGroupAndName().

pfrenssen committed:
Fix PHP CodeSniffer warnings.

pfrenssen commented:
I'm ready with this! I have addressed all the remarks (I think :) )

amitaibu commented on src/Og.php:
```diff
@@ -604,23 +606,6 @@ public static function groupManager() {
-   * @return \Drupal\og\OgRoleInterface|null
-   *   The OG role object, or NULL if a matching role was not found.
-   */
-  public static function getRole($entity_type_id, $bundle, $role_name) {
```
As indicated in another comment getRole should still remain (although we can move it to OgRole itself ) as You may want to set the role's permissions [before](https://github.com/amitaibu/og/blob/8.x-1.x/tests/src/Functional/GroupSubscribeTest.php#L139-L143) any group as actually created

Also, just as a heads up, the entire OgRole will still undergo changes, once we'll have to introduce the concept of `OgRolePerBundle` and `OgRolePerGroup` (as we have in OG7)

pfrenssen replied:
OK will restore it! I will make a followup to move it to `OgRole`, because it will need some discussion about the correct naming of the function. Burying this in here will cause it to be missed by everyone who is involved in OG atm.

pfrenssen replied:
Followup: https://github.com/amitaibu/og/issues/292

amitaibu commented on src/OgAccess.php:
```diff
@@ -253,6 +281,86 @@ public function userAccessEntity($operation, EntityInterface $entity, AccountInt
+      ];
+      $membership = Og::getMembership($group_entity, $user, $states) ?: $this->getNonMemberMembership($group_entity, $user);
+      foreach ($permissions as $permission) {
+        if ($membership->hasPermission($permission->getName())) {
```
This non-member membership is problematic IMO. Apart of the fact that nothing prevents it form being saved, it feels conceptually wrong. If you an OgMembership == you are an (active|pending|blocked) _member_.

We could use $og_access->userAccess, without the need for creating dummy memberships.

pfrenssen replied:
I did this following your recommendation in https://github.com/amitaibu/og/pull/217#discussion_r73567630 - I had split off the creation of the membership to a separate method for readability. I made that method protected to prevent people from calling it and saving the membership that it returns.

_Anyway_ you are completely correct, we totally need to replace it with a call to `::userAccess()` here, because otherwise the alter hooks would not fire! I somehow missed that little detail :confounded:

pfrenssen replied:
I've implemented it and it works fine for everything except blocked users. Inside `OgAccess::userAccess()` blocked users are currently ignored when the membership is retrieved. This causes blocked users to be treated as non-members. Yay for the tests discovering bugs! Will fix.

amitaibu commented on src/OgAccess.php:
```diff
@@ -310,4 +418,33 @@ public function reset() {
+      ->setUser($user)
+      ->setGroup($group)
+      ->setState(OgMembershipInterface::STATE_PENDING)
+      ->setRoles([new OgRole(['id' => $role_id])]);
```
This is probably related to https://github.com/amitaibu/og/pull/217/files#r73741176 - For me it looks wrong. We are creating a lot of dummy data, just to get the access, which we are already doing [here](https://github.com/amitaibu/og/blob/8.x-1.x/src/OgAccess.php#L155-L160)

pfrenssen replied:
Right! Got rid of it :)

amitaibu commented:
Thanks for keeping up the good work!

Made comments about the "dummy non-member memberships" (dnmm in short 😜).
(Also it will need a re-roll)

pfrenssen committed:
Leverage OgAccess::userAccess() instead of duplicating half its logic and ignoring the other half.

pfrenssen committed:
Add documentation.

pfrenssen committed:
Remove unused variable.

pfrenssen committed:
Blocked users should not be granted any permissions.

pfrenssen committed:
Revert "Provide a convenient method OgRole::loadByGroupAndName()."

This partially reverts commit 82f016fa28d6d63fefa6d0135309156618c3f4f7.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into check-group-content-crud-access

pfrenssen committed:
Fix PHP_CodeSniffer warnings.

pfrenssen commented:
Addressed the remarks, rerolled, and tests are green :)

amitaibu commented on src/OgAccess.php:
```diff
@@ -140,22 +166,30 @@ public function userAccess(EntityInterface $group, $operation, AccountInterface
+        OgMembershipInterface::STATE_BLOCKED,
+      ];
+      if ($membership = Og::getMembership($group, $user, $states)) {
+        // Blocked users don't have any permissions.
```
👍

amitaibu commented on src/OgAccess.php:
```diff
@@ -140,22 +166,30 @@ public function userAccess(EntityInterface $group, $operation, AccountInterface
         // User is a non-member.
         /** @var \Drupal\og\Entity\OgRole $role */
-        $role = Og::getRole($group_type_id, $bundle, OgRoleInterface::ANONYMOUS);
+        $role = OgRole::loadByGroupAndName($group, OgRoleInterface::ANONYMOUS);
```
👍 the name is fine for now - things might shift a bit once we'll have the perBundle/ perGroup roles.

amitaibu commented:
Wonderful, thanks!

pfrenssen commented:
Woohoo!! Thanks a lot!!
